In [1]:
import pandas as pd 
import numpy as np
from sklearn import preprocessing
data=pd.read_csv("final_data.csv")
#final_data is the final pre processed data

In [2]:
data.head()

Operator In Out Travelling Network Type  Rating Call Drop Category  \
0    Airtel            Indoor           4G       3       Satisfactory   
1  Vodafone            Indoor           4G       1       Call Dropped   
2    Airtel            Indoor           3G       1       Call Dropped   
3    Airtel            Indoor           4G       3       Satisfactory   
4    Airtel        Travelling           4G       3       Satisfactory   

   Latitude  Longitude   State Name  Average Data Speed Download  \
0  0.196228   0.325356    Karnataka                     0.271480   
1  0.632907   0.135432      Gujarat                     0.211519   
2  0.127093   0.290671   Tamil Nadu                     0.036321   
3  0.580181   0.155327      Gujarat                     0.175517   
4  0.425737   0.180503  Maharashtra                     0.333021   

   Average Data Speed Upload  Population  Literacy Rate  Rainfall  \
0                   0.197512           2              7         3   
1                   0.370879           2              8         2   
2                   0.062303           2              8         2   
3                   0.191200           2              8         2   
4                   0.395612           4              8         4   

   Area of the state  Average_Signal  
0           0.504087        0.476721  
1           0.518040        0.477252  
2           0.300607        0.478454  
3           0.518040        0.477699  
4           0.886194        0.476859

In [3]:
Vodafone=data[data['Operator']=='Vodafone']

In [4]:
Vodafone.count()

Operator                       5895
In Out Travelling              5895
Network Type                   5895
Rating                         5895
Call Drop Category             5895
Latitude                       5895
Longitude                      5895
State Name                     5895
Average Data Speed Download    5895
Average Data Speed Upload      5895
Population                     5895
Literacy Rate                  5895
Rainfall                       5895
Area of the state              5895
Average_Signal                 5895
dtype: int64

In [5]:
from sklearn import preprocessing
from numpy.linalg import inv
from sklearn.linear_model import LinearRegression

In [6]:
X=Vodafone[['Average Data Speed Download','Average Data Speed Upload']]

In [7]:
X.count()

Average Data Speed Download    5895
Average Data Speed Upload      5895
dtype: int64

In [8]:
y=Vodafone['Average_Signal']

In [10]:
from sklearn.model_selection import train_test_split 
x_train, x_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1) 
x_train.reset_index(inplace = True, drop = True) 
x_test.reset_index(inplace = True, drop = True) 
y_train.reset_index(inplace = True, drop = True) 
y_test.reset_index(inplace = True, drop = True) 
#We have used holdout method and 70%of data is taken as training data and rest as testing

In [11]:
x_train.head()

Average Data Speed Download  Average Data Speed Upload
0                     0.058113                   0.079079
1                     0.069154                   0.098923
2                     0.175813                   0.397521
3                     0.068401                   0.081367
4                     0.147096                   0.237023

In [12]:
n=x_train.count()[0]

In [13]:
n

4126

# Regression Model of degree 1

In [14]:
sum_x1=0
sum_x2=0
sum_y1=0
sum_x1x1=0
sum_x1x2=0
sum_x2x2=0
sum_x1y1=0
sum_x2y1=0
for i in range(0,n):
    sum_x1=sum_x1+x_train['Average Data Speed Download'][i]
    sum_x2=sum_x2+x_train['Average Data Speed Upload'][i]
    sum_y1=sum_y1+y_train[i]
    sum_x1x1=sum_x1x1+x_train['Average Data Speed Download'][i]*x_train['Average Data Speed Download'][i]
    sum_x1x2=sum_x1x2+x_train['Average Data Speed Download'][i]*x_train['Average Data Speed Upload'][i]
    sum_x2x2=sum_x2x2+x_train['Average Data Speed Upload'][i]*x_train['Average Data Speed Upload'][i]
    sum_x1y1=sum_x1y1+x_train['Average Data Speed Download'][i]*y_train[i]
    sum_x2y1=sum_x2y1+x_train['Average Data Speed Upload'][i]*y_train[i]
A = np.array([[n, sum_x1, sum_x2], [sum_x1, sum_x1x1, sum_x1x2],[sum_x2,sum_x1x2,sum_x2x2]])
B=np.array([sum_y1, sum_x1y1, sum_x2y1])
#All the sums are used to calculate A and then X is calculated using AX=B so X=Inverse(A)*B 
#X will give all the parameters(w0,w1,w2)

In [15]:
A

array([[4126.        ,  540.00115372,  931.97119014],
       [ 540.00115372,   82.34933271,  148.80874202],
       [ 931.97119014,  148.80874202,  281.56703098]])

In [16]:
B

array([1971.03464538,  257.79342396,  444.80293844])

In [17]:
Inverse=inv(A)
w0=Inverse[0][0]*B[0]+Inverse[0][1]*B[1]+Inverse[0][2]*B[2]
w1=Inverse[1][0]*B[0]+Inverse[1][1]*B[1]+Inverse[1][2]*B[2]
w2=Inverse[2][0]*B[0]+Inverse[2][1]*B[1]+Inverse[2][2]*B[2]
W=np.array([w0,w1,w2])

In [18]:
W

array([ 0.47949249, -0.01051505, -0.00179538])

In [19]:
y_pred=w0+w1*x_test['Average Data Speed Download']+w2*x_test['Average Data Speed Upload']
#y_pred gives the predicted values of y based on the regression model
MSE=0
for i in range(0,len(y_test)):
    MSE=MSE+(y_pred[i]-y_test[i])**2
RMSE1=np.sqrt(MSE/len(y_test))
RMSE1
#RMSE is used to find the goodness of the model and is calacuated as root of sum of square of difference of actual and prediacted values.

0.0008605484822797491

In [20]:
ybar=y_test.mean()
ybar

0.47774244744431227

In [22]:
SSR=0
SST=0
SSE=0
y_test.count()
#SST is total sum of squares
#SSR is sum of squares due to regression
#SSE is sum of squares due to error

1769

In [23]:
for i in range (0,len(y_test)):
    SST=SST+(y_test[i]-ybar)*(y_test[i]-ybar)
    SSR=SSR+(y_pred[i]-ybar)*(y_pred[i]-ybar)
    SSE=SSE+(y_test[i]-y_pred[i])*(y_test[i]-y_pred[i])    

In [24]:
print("Total sum of squares=", SST )

Total sum of squares= 0.0023771228059470757


In [25]:
print("Sum of squares due to regression =", SSR)

Sum of squares due to regression = 0.0010536112111918076


In [26]:
print("Sum of square due to error=", SSE)

Sum of square due to error= 0.0013100217882361898


In [27]:
Rsquare=SSR/SST

In [28]:
print("R square value=",Rsquare)

R square value= 0.4432296087336706


Since there are 2 independent variable so R square adjusted is also calculated. 

In [29]:
p=2
n=len(y_test)


In [30]:
Rsquare_adjusted=1-(((1-Rsquare)*(n-1))/(n-p-1))

In [31]:
Rsquare_adjusted

0.4425990646892015

Significance Test

F Test


In [32]:
W

array([ 0.47949249, -0.01051505, -0.00179538])

In [33]:
w0=W[0]
w0

0.4794924930921072

In [34]:
w1=W[1]
w2=W[2]


Regression line is y= bo + b1 * x1 + b2 * x2

b1 is coefficient of x1

b2 is coefficient of x2

H0 : b1=b2=0

Ha : One or both parameters not equal to zero

In [35]:
MSR=SSR/2
MSR

0.0005268056055959038

In [36]:
MSE=SSE/(len(y_test)-3)
MSE

7.418016920929727e-07

In [45]:
F=MSR/MSE
#The F test is used to determine whether a significant relationship exists between the dependent variable and the set of all the independent variables; hence F test are referred to as  the test for overall significance

In [46]:
F

710.1704016197867

In [47]:
len(y_test)-3

1766

F alpha with numerator dof=2 and denominator dof=n-3(=1766) at level of significance=0.05 is equal to 3.0008

In [48]:
Falpha=3.00081980

In [49]:
F>=Falpha

True

So we reject H0. Now for b1 and b2 t tests are performed separately.


T test for b1

H0 : b1=0

Ha : b1 is not equal to zero

In [50]:
sum=0
m=x_test['Average Data Speed Download'].mean()
for i in range(0,len(x_test)):
    sum=sum+((x_test['Average Data Speed Download'][i]-m)**2)
sum

4.897642655768034

In [51]:
import math
x=math.sqrt(sum)
z=SSE/(len(y_test)-3)
s=math.sqrt(z)
sb1=s/x

In [52]:
t=w1/sb1

In [53]:
t
#t distribution with n − p − 1 degrees of freedom.
#The t test is used to determine whether each of the individual independent variables is significant. A separate t test is conducted for each of the independent variables in the model; hence t tests are called as the test for individual significance. 

-27.018490538490138

In [54]:
len(y_test)-3

1766

So talpha=2.243316 at dof=len(y_test)-3 and alpha=0.025

In [55]:
talpha=2.243316
t<=-talpha

True

Hence we reject null hypothesis

T Test for b2

H0 : b2=0

Ha : b2 not equal to zero

In [56]:
sum=0
m=x_test['Average Data Speed Upload'].mean()
for i in range(0,len(x_test)):
    sum=sum+((x_test['Average Data Speed Upload'][i]-m)**2)
x=math.sqrt(sum)
z=SSE/(len(y_test)-3)
s=math.sqrt(z)
sb2=s/x

In [57]:
t=w2/sb2

In [58]:
t

-11.244066284326527

In [59]:
t<=-talpha

True

Hence we reject null hypothesis

## So we can conclude that Signal Strength does depend on Download and Upload Speed significantly for Airtel.